File: Single_source_shortest_paths/BellmanFord's_algorithm/bellmanford.py
def bellmanford(adj_list, start_vertex, tw, noe):
    
    distance = [tw for i in range(len(adj_list))]
   
    distance[start_vertex] = 0
    
    while(noe):
    
        for i in range(len(adj_list)):
        
            for key, value in adj_list[i].items():
        
                distance[key] = min(distance[key], (distance[i] + value))
                             
        
        noe -= 1
        
    vertex_no = 0        
    
    for each in distance:
        
        print("Shortest distance to vertex", vertex_no," : ", each)
        
        vertex_no += 1 


def check_vertex(vertex, nov):
    
    while(1):
        
        if (vertex > nov or vertex < 0):
            
            print("Invalid vertex: ", vertex)
            
            vertex = int(input("Re-enter the vertex: "))
        
        else:
            return vertex
            
            
def read_edges(nov, noe, adj_list):                  # noe = no_of_edges
    
    tw = 1
    
    print("\nProvide the edges and their weights: \nvertex1 vertex2 weight", end="\n")
    
    while (noe):

        vertex = list(input().split(" "))
        source, destination, weight = int(vertex[0]), int(vertex[1]), int(vertex[2])     
    
        source = check_vertex(source, nov)
        destination = check_vertex(destination, nov)
        
        tw += abs(weight)
        
        adj_list[source][destination] = weight
        
        noe -= 1
    
    return (tw)         

###################################################

no_of_vertices = int(input("Enter the no of vertices: "))

adj_list = [{} for i in range(0, (no_of_vertices))]

no_of_edges = int(input("\nEnter the no of edges: "))    

total_weight = read_edges(no_of_vertices, no_of_edges, adj_list)

start_vertex = int(input("\nEnter the start vertex: "))

print("\nCalling BellmanFord's algorithm...", end="\n")

bellmanford(adj_list, start_vertex, total_weight, no_of_edges)

File: Single_source_shortest_paths/BellmanFord's_algorithm/bellmanford.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>


void bellmanford(std::vector<std::map<int,int>> adj, int& start, int& twc, int vc)
{

	int initial_vertex = start;

	std::vector<int> distance(adj.size(), twc);

	distance[initial_vertex] = 0;

	while(vc--)        // loop will run vertices_count - 1 times
	{

		for(int i = 0; i < adj.size(); i++)
		{
			
      for(auto neighbour : adj[i])
			{
       
        distance[neighbour.first] =  std::min(distance[neighbour.first], distance[i] + neighbour.second);
      
      }

    }
	
  }
	
  int count = 0;

	for(auto& d : distance)
	{
		std::cout << "Distance from "<< initial_vertex << " to  vertex " << count << " is: "<< d << std::endl;
		count++;
	}


}


void check_vertex(int& v, int& vc)	// check_vertex starts
{
	while (v >= vc || v < 0)
	{
		std::cout << "Invalid vertex no: " << v << std::endl;
		std::cout << "Re enter the vertex: " << std::endl;
		std::cin >> v;
	}
}	// check_vertex ends


void read_vertex_and_edge_count(int& v_count, int& e_count)
{

	std::cout << "Enter the no of vertices:" << std::endl;
	std::cin >> v_count;
	std::cout << "Enter the no of edges: " << std::endl;
	std::cin >> e_count;

}


void read_the_edges(std::vector<std::map<int,int>>& adj, int ec, int& v_count, int& twc)
{

	int source;
	int destination;
	int weight;
	std::cout << "\nEnter the edges and their weights: " << std::endl;

	while (ec--)
	{
		std::cin >> source >> destination >> weight;

		check_vertex(source, v_count);
		check_vertex(destination, v_count);

		weight < 0 ? twc += ((-1) * (weight)) : twc += weight;         // Representing the concept of infinity value at the beginning  

		adj[source].insert({destination, weight});

	}

}


int main()
{

	int start_vertex;
	int vertex;
	int vertices_count = 0;
	int edges_count = 0;
	int total_weight_count = 2;
	read_vertex_and_edge_count(vertices_count, edges_count);

	std::vector<std::map<int,int>> adjacency_list(vertices_count);

	read_the_edges(adjacency_list, edges_count, vertices_count, total_weight_count);

	std::cout << "\nEnter the start vertex: ";
	std::cin >> start_vertex;

	check_vertex(start_vertex, vertices_count);

	bellmanford(adjacency_list, start_vertex, total_weight_count, (vertices_count - 1));             // CALLING THE BELLMAN-FORD's ALGORITHM

	return 0;
}
